libs/utils/analysis/idle_analysis.py

```diff
@@ -0,0 +1,266 @@
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+""" Idle Analysis Module """
+
+import matplotlib.gridspec as gridspec
+import matplotlib.pyplot as plt
+import pandas as pd
+import pylab as pl
+
+from analysis_module import AnalysisModule
+from trace import ResidencyTime, ResidencyData
+from trappy.utils import listify
+
+# Configure logging
+import logging
+
+
+class IdleAnalysis(AnalysisModule):
+    """
+    Support for plotting Idle Analysis data
+
+    :param trace: input Trace object
+    :type trace: :mod:`libs.utils.Trace`
+    """
+
+    def __init__(self, trace):
+        super(IdleAnalysis, self).__init__(trace)
+
+###############################################################################
+# DataFrame Getter Methods
+###############################################################################
+
+    def _dfg_idle_state_residency(self, entity):
+        """
+        Compute time spent by a given entity in each idle state.
+
+        :param entity: cpu ID or cluster name or list of CPU IDs
+        :type entity: int or str
+
+        :returns: :mod:`pandas.DataFrame` - idle state residency dataframe
+        """
+        if not self._trace.hasEvents('cpu_idle'):
+            logging.warn('Events [cpu_idle] not found, '\
+                         'idle state residency computation not possible!')
+            return None
+
+        if isinstance(entity, str):
+            try:
+                _entity = self._platform['clusters'][entity.lower()]
+            except KeyError:
+                logging.warn('%s entity not found!', entity)
+                return None
+        else:
+            _entity = listify(entity)
+
+        idle_df = self._dfg_trace_event('cpu_idle')
+        entity_idle_df = idle_df[idle_df.cpu_id == _entity[0]]
+
+        # Build cpu_idle, a square wave of the form:
+        #     entity_idle[t] == 1 if all CPUs in the entity are reported to be
+        #                       idle by cpufreq at time t
+        #     entity_idle[t] == 0 otherwise
+        entity_idle = self._trace.getClusterActiveSignal(_entity) ^ 1
+
+        # In order to compute the time spent in each idle statefrequency we
+        # multiply 2 square waves:
+        # - entity_idle
+        # - idle_state, square wave of the form:
+        #     idle_state[t] == 1 if at time t entity is in idle state i
+        #     idle_state[t] == 0 otherwise
+        available_idles = sorted(idle_df.state.unique())
+        # Remove non-idle state from availables
+        available_idles.pop()
+        new_idx = sorted(entity_idle_df.index.tolist() + \
+                         entity_idle.index.tolist())
+        entity_idle_df = entity_idle_df.reindex(new_idx, method='ffill')
```

Comment:
This one seems to explode with

> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> <ipython-input-9-564585c37313> in <module>()
> ----> 1 trace.analysis.idle.plotCPUIdleStateResidency()
> 
> /media/hdd1tb/work/lisa/libs/utils/analysis/idle_analysis.pyc in plotCPUIdleStateResidency(self, cpus, pct)
>     144         xmax = 0.0
>     145         for cpu in _cpus:
> --> 146             r = self._dfg_idle_state_residency(cpu)
>     147             residencies.append(ResidencyData('CPU{}'.format(cpu), r))
>     148 
> 
> /media/hdd1tb/work/lisa/libs/utils/analysis/idle_analysis.pyc in _dfg_idle_state_residency(self, entity)
>      89         new_idx = sorted(entity_idle_df.index.tolist() + \
>      90                          entity_idle.index.tolist())
> ---> 91         entity_idle_df = entity_idle_df.reindex(new_idx, method='ffill')
>      92         entity_idle = entity_idle.reindex(new_idx, method='ffill')
>      93         idle_time = []
> 
> /usr/local/lib/python2.7/dist-packages/pandas/core/frame.pyc in reindex(self, index, columns, **kwargs)
>    2682     def reindex(self, index=None, columns=None, **kwargs):
>    2683         return super(DataFrame, self).reindex(index=index, columns=columns,
> -> 2684                                               **kwargs)
>    2685 
>    2686     @Appender(_shared_docs['reindex_axis'] % _shared_doc_kwargs)
> 
> /usr/local/lib/python2.7/dist-packages/pandas/core/generic.pyc in reindex(self, *args, **kwargs)
>    1960         # perform the reindex on the axes
>    1961         return self._reindex_axes(axes, level, limit, tolerance,
> -> 1962                                   method, fill_value, copy).__finalize__(self)
>    1963 
>    1964     def _reindex_axes(self, axes, level, limit, tolerance, method,
> 
> /usr/local/lib/python2.7/dist-packages/pandas/core/frame.pyc in _reindex_axes(self, axes, level, limit, tolerance, method, fill_value, copy)
>    2631         if index is not None:
>    2632             frame = frame._reindex_index(index, method, copy, level,
> -> 2633                                          fill_value, limit, tolerance)
>    2634 
>    2635         return frame
> 
> /usr/local/lib/python2.7/dist-packages/pandas/core/frame.pyc in _reindex_index(self, new_index, method, copy, level, fill_value, limit, tolerance)
>    2639         new_index, indexer = self.index.reindex(new_index, method, level,
>    2640                                                 limit=limit,
> -> 2641                                                 tolerance=tolerance)
>    2642         return self._reindex_with_indexers({0: [new_index, indexer]},
>    2643                                            copy=copy, fill_value=fill_value,
> 
> /usr/local/lib/python2.7/dist-packages/pandas/core/index.pyc in reindex(self, target, method, level, limit, tolerance)
>    2131                 else:
>    2132                     if method is not None or limit is not None:
> -> 2133                         raise ValueError("cannot reindex a non-unique index "
>    2134                                          "with a method or limit")
>    2135                     indexer, missing = self.get_indexer_non_unique(target)
> 
> ValueError: cannot reindex a non-unique index with a method or limit

but it might be a trace related issue.